README.md
# Validate Fields
[![Build Status](https://travis-ci.org/clubedaentrega/validate-fields.svg?branch=master)](https://travis-ci.org/clubedaentrega/validate-fields)
[![Inline docs](http://inch-ci.org/github/clubedaentrega/validate-fields.svg?branch=master)](http://inch-ci.org/github/clubedaentrega/validate-fields)
[![Dependency Status](https://david-dm.org/clubedaentrega/validate-fields.svg)](https://david-dm.org/clubedaentrega/validate-fields)

A simple yet powerful JSON schema validator

## Install
`npm install validate-fields`

## Usage

### Basic
```javascript
let validate = require('validate-fields')(),
	schema = {
		name: String,
		age: 'uint'
	}, value = {
		name: 'Ann',
		age: 27
	}

// Validate the value against the schema definition
if (validate(schema, value)) {
	console.log('Valid!')
} else {
	console.log('Invalid', validate.lastError)
}
```
`validate` throws if it could not understand the schema and returns false if the schema is ok but the value doesn't match the schema.

`validate.lastError`, `validate.lastErrorMessage` and `validate.lastErrorPath` will let you know about the validation error cause

See the list of valid values for the schema bellow

### Pre-parsing
```javascript
// Parse the schema once and store it
let schema = validate.parse([String]) // an array of non-empty strings

// Set express route
// app here is an express instance, used only to ilustrate this example (not part of this module!)
app.post('/api/posts/by-tags', function (req, res, next) {
	// Validate using the parsed schema
	if (schema.validate(req.body)) {
		next()
	} else {
		// Answer with error (note that we're not using validate.lastError here)
		res.status(400).send(schema.lastError)
	}
}, handlePostsByTags)
```
This way is recommended because if the schema definition is invalid, `validate.parse()` will throw early and `schema.validate()` won't throw. You also gain in speed!

## Escaping HTML
HTML strings can be escaped by sending one more argument to `validate`:
```javascript
let v = ['html>']
validate([String], v)
v[0] === 'html>'

validate([String], v, {
	escape: true
})
v[0] === 'html&gt;'
```
Or, using pre-parsing
```javascript
let v = ['html>']
validate.parse([String]).validate(v, {
	escape: true
})
v[0] === 'html&gt;'
```

## Strict matching
By default, extra keys are not considered errors:
```javascript
validate({}, {a: 2}) // true
```

This can be changed with the `strict` option:
```javascript
validate({}, {a: 2}, {strict: true}) // false
```

## Standard types

### Hash map (object)
Example: `{a: Number, 'b?': String, c: {d: 'int'}}`

Keys that end in `'?'` are optional (`b`, in the example above). All others must be present and must not empty. Empty is `null` or `undefined`; for strings, `''` is also considered empty.

If a value is optional and is empty it will be removed. Example:
```javascript
let obj = {a: '', b: null}
validate({'a?': String, 'b?': 'int'}, obj) // true
obj // {}
```

Optional fields may declare a default value to use in its place when empty. Example:
```js
let obj = {}
validate({'a=12': Number, 'b=[]': [String], 'c="hi"': String}, obj) // true
obj // {a: 12, b: [], c: 'hi'}
```

The value after the `=` character must be a valid JSON.

### Array
Example: `{books: [{title: String, author: String}]}`

### Other types
* `Number`: a double
* `String`: a string
* `Object`: any non-null object
* `Array`: any array
* `Boolean`
* `Date`: any date string accepted by Date constructor (ISO strings are better though)
* `'*'`: anything
* `'number(-3.5,10)'`: a number `x` with `-3.5 <= x <= 10`
* `'number(-3.5,)'`: a number `x` with `-3.5 <= x`
* `'number(,10)'`: a number `x` with `x <= 10`
* `'numeric'`: a double value encoded in a string
* `'numeric(<...>)'`: like `'number(<...>)'`, but for numeric strings
* `'int'`: an integer between -2^51 and 2^51 (safe integer)
* `'int(-3,10)'`: an integer between -3 and 10. Lower and upper bounds are optional
* `'numericInt'`: like `'int'`, but for numeric strings
* `'uint'`: a natural number less than 2^51
* `'uint(3,10)'`: a natural number between 3 and 10. Lower and upper bounds are optional
* `'numericUint'`: like `'int'`, but for numeric strings
* `'string(17)'`: a string with exactly 17 chars
* `'string(,100)'`: at most 100 chars
* `'string(8,)'`: at least 8 chars
* `'string(8,100)'`: at most 100, at least 8
* `'hex'`: a hex string (even number of hex-chars)
* `'hex(12)'`: a hex-string with exactly 12 hex-chars (that is, 6 bytes)
* `'base64'`: a base64 string, with valid padding
* `'id'`: a mongo objectId as a 24-hex-char string
* `'email'`: a string that looks like an email address
* `'in(cat, dog, cow)'`: a string in the given set of strings
* `'numberIn(3, 1.4, -15)'`: a number in the given set of values
* `'numericIn(<...>)'`: like `'numberIn(<...>)'`, but for numeric strings
* `/my-own-regex/`: a string that matches the custom regexp

## Custom types
The simplest way to define your own type is like this:
```javascript
validate.typedef('name', {
	first: String,
	'last?': String
})
validate.typedef('zip-code', /^\d{5}([- ]\d{4})?$/)

validate('name', {last:'Souza'}) // false
validate.lastError // 'I was expecting a value in first'
validate(['zip-code'], ['12345', '12345-1234']) // true
```
Typedef works as a simple alias (like in C)

Every time you call `require('validate-fields')()` a new context is created. Each context is isolated from each other. Types defined in one context do not conflict with other contexts.

If you need more power, you can create the type from scratch and define every detail about the validation.

In the example bellow, we create a type defined by 'divBy3' that matches JSON numbers.
The third param is the check function, that should throw when the value is invalid.
It may return a new value, that will replace the original value in the object
```javascript
validate.registerType('divBy3', 'number', function (value) {
	if (value%3 !== 0) {
		// Note: you should throw a string
		// If you throw an Error instance it won't be caught!
		throw 'I was expecting a number divisible by 3'
	}
	return value/3
})

let obj = {n: 12}
validate({n: 'divBy3'}, obj) // true
obj.n // 4
```

In the example bellow, we implement the general case of the above. Instead of a fixed string (divBy3) we define a syntax like `'divBy(n)'` where `n` is a number. The "argument" list is sent as the third parameter to the check function.
```javascript
validate.registerTaggedType({
	tag: 'divBy',
    jsonType: 'number',
    minArgs: 1, // default = 0. If 0, 'tag' and 'tag()' are equal
    maxArgs: 1, // default = 0 = no limit
    sparse: false, // default = false. If true, let arguments be skipped: 'tag(1,,2)'
    numeric: true // default = false. If true, parse all arguments as numbers
}, function (value, args) {
	let n = args[0]
	if (value%n !== 0) {
		throw 'I was expecting a number divisible by '+n
	}
	return value/n
})
validate('divBy(17)', 35) // false
validate('divBy(35)', 35) // true
```
See more examples in the folder `types`. All core types are defined there

Finally, you can extend a type by providing pre and post hooks:
```javascript
validate.typedef('time-in', {
	hour: /^\d\d$/,
	minute: /^\d\d$/
}, null, value => {
	// Convert that object to a string like 'HH:MM'
	return value.hour + ':' + value.minute
})

validate.typedef('time-out', {
	hour: /^\d\d$/,
	minute: /^\d\d$/
}, value => ({
	// Convert 'HH:MM' to an object
	hour: value.slice(0, 2),
	minute: value.slice(3, 5)
}))
```

Pre-hooks will be executed before the base type checks. Post-hooks will be run afterwards.

## Interchangeable format
A parsed `Field` can be serialized to JSON with `JSON.stringify()`:
```javascript
let fields = validate.parse({name: String, age: 'uint', 'birth?': Date})
JSON.stringify(fields) // '{"name":"$String","age":"uint","birth?":"$Date"}'
```

Note that custom types can't be serialized, they are replaced by their JSON-type:
```javascript
let fields = validate.parse({myType: 'divBy(7)'}) // defined above
JSON.stringify(fields) // '{"myType":"$Number"}'
```

To get back a `Field` instance, `JSON.parse()` should be used with a reviver and the result then parsed:
```javascript
let serializedFields = '{"name":"$String","age":"uint","birth?":"$Date"}'
let definition = JSON.parse(serializedFields, validate.reviver) // {name: String, age: 'uint', 'birth?': Date}
let fields = validate.parse(definition)

fields.validate({name: 'John', age: 12}) // true
```

## JSON Schema
A parsed `Field` can be exported as [JSON Schema](http://json-schema.org/):
```javascript
let fields = validate.parse({name: String, age: 'uint', 'birth?': Date})
fields.toJSONSchema()
/* {
	type: 'object',
	properties: {
		name: {
			type: 'string'
		},
		age: {
			type: 'integer',
			minimum: 0
		},
		birth: {
			type: 'string',
			format: 'date-time'
		}
	},
	required: ['name', 'age']
} */
```

## Partial validation
A schema can be used for validating only a subset of fields, for example:

```javascript
let schema = validate.parse({label: String, creation: Date})
schema.validate({label: 'earth'}) // false
schema.validate({label: 'earth'}, {partial: ['label']}) // true
```

This is useful when fields in the value can be dynamically selected, like in an endpoint when the caller only cares about two or three fields from a normally bigger schema.

Multiple paths should be passed as elements of the `options.partial` array. Paths can also recurse into inner properties:

```javascript
validate({
	post: {title: String, creation: Date},
	user: {name: String, age: 'uint'}
}, {
	post: {title: 'Top 10 ways to create top listings'},
	user: {age: 30}
}, {
	partial: ['post.title', 'user.age']
}) // true
```

In arrays, partial paths enumerate over its elements:

```javascript
validate({
	table: {
		rows: [{
			cells: [{value: Number, text: String}]
		}]
	}
}, {
	table: {
		rows: [{
			cells: [{value: 3}, {value: 13}]
		}, {
			cells: [{value: 14}, {value: '15'}]
		}]
	}
}, {
	partial: ['table.rows.cells.value']
}) // false
validate.lastError // I was expecting number and you gave me string in table.rows.1.cells.1.value
```

Custom values should implement partial validation accordingly, using the value received over `options._partialTree`. Example:

```javascript
validate.registerType('custom-with-partial', 'object', (value, extra, options) => {
	options._partialTree // {c: {d: false}}
})
validate(
	{a: {b: 'custom-with-partial'}},
	{a: {b: 'some data'},
	{partial: ['a.b.c.d']}
)
```

In strict mode, any field in the value not listed in the partial will result in an error.
Likewise, any field in the partial not in the schema will result in an error.

HISTORY.md
# 5.1.1
* Fixed: `strict` option was not totally respected when using `partial`

# 5.1.0
* Added: option `partial` to allow for partial validation

# 5.0.0
## Breaking changes
* `Field#toJSONSchema()` when converting a type defined with typedef now behaves differently.
Now it returns the full description when called with no paramenters.
To use references, pass the base path as the first parameter, example: `Field#toJSONSchema('#/definitions')`

# 4.1.0
* Fix: apply pre hook before checking for emptiness

# 4.0.0
## Breaking changes
* Drop support for node 0.12 and 4

## Other changes
* Added: pre and post hooks to `typedef()`

# 3.7.1
* Changed: do not depend on Buffer to check base64 string

# 3.7.0
* Added: `parseArgs` callback for `registerTaggedType`, called to pre-process tag arguments into extra

# 3.6.1
* Fixed: use isFinite instead of Number.isFinite to support IE

# 3.6.0
* Added: `context.registerObjectType` to be used by core types in order to support IE

# 3.5.1
* Fixed: use isNaN instead of Number.isNaN to support IE

# 3.5.0
* Added: `Field#toJSONSchema()` that returns a parsed schema encoded in the [JSON Schema](http://json-schema.org/) standard

# 3.4.0
* Added: `'numeric'`, `'numericInt'`, `'numericUint'` and `'numericIn'` core types. Those are string equivalents of `Number`, `'int'`, `'uint'`, `'numberIn'` types

# 3.3.1
* Fixed: NaN, Infinity and -Infinity being accepted as numbers [issue #11](https://github.com/clubedaentrega/validate-fields/issues/11)

# 3.3.0
* Added: `'base64'`
* Fixed: `'hex'` does not accept an odd number of bytes

# 3.2.1
* Fixed: message for `number(min,max)`, `int(min,max)`, `uint(min,max)`

# 3.2.0
* Added: json type `'raw'` (like `'*'`, but don't call toJSON)

# 3.1.0
* Added: remove extraneous keys with `undefined` value in strict mode. Now `validate({}, {a: undefined}, {strict: true})` is `true`

# 3.0.0
## Breaking changes
* `validate({a: [String]}, {a: []})` now returns `true`

## Issues closed
*  Changed: do not consider empty array as an empty value [issue #9](https://github.com/clubedaentrega/validate-fields/issues/9) [pull #10](https://github.com/clubedaentrega/validate-fields/pull/10)

# 2.1.0
* Changed: `toJSON` is called when present, so now `validate(new Date, Date)` is `true`.

# 2.0.0
## Breaking changes
* You must use `var validate = require('validate-fields')()` instead of old `var validate = require('validate-fields')` (note the `()` at the end).
* `validate(String, '')` now returns `true` instead of `false`
* `validate({'a?': Number}, {a: ''})` now returns `false` instead of `true`

## Issues closed
* Added: default values [issue #3](https://github.com/clubedaentrega/validate-fields/issues/3)
* Changed: new definition of emptiness [issue #2](https://github.com/clubedaentrega/validate-fields/issues/2) [pull #5](https://github.com/clubedaentrega/validate-fields/pull/5)
* Changed: removed empty check from other places than hash [issue #8](https://github.com/clubedaentrega/validate-fields/issues/8)
* Changed: no more global registry, create contexts instead [issue #7](https://github.com/clubedaentrega/validate-fields/issues/7)

# 1.4.0
* Added: new numeric types [issue #1](https://github.com/clubedaentrega/validate-fields/issues/1)
* Fixed: array is accepted as object [issue #4](https://github.com/clubedaentrega/validate-fields/issues/4)
* Added: give support for tagged types [issue #6](https://github.com/clubedaentrega/validate-fields/issues/6)

# 1.3.0
* Added: better error reporting (lastErrorPath, lastErrorMessage)
* Added: type '*'
* Added: strict option

# 1.2.0
* Added: toJSON and from JSON support
* Fixed: to exact-length string rule
* Fixed: to null-prototyped objects

types.js
'use strict'

var register = require('./index.js').registerType,
	registerTagged = require('./index.js').registerTaggedType,
	isEmpty = require('./Type.js').isEmpty,
	ValidationError = require('./ValidationError')

/**
 * @param {string} str
 * @returns {string}
 */
function escape(str) {
	return str.replace(/&/g, '&amp;')
		.replace(/</g, '&lt;')
		.replace(/>/g, '&gt;')
		.replace(/'/g, '&#39;')
		.replace(/"/g, '&quot;')
}

/**
 * Hash maps
 * Example: {name: String, age: Number, dates: {birth: Date, 'death?': Date}}
 * If the key name ends with '?' that field is optional
 * After validation, a field will never be empty (like '' or []). If it's marked as optional, it will be either non-empty or not-present ('key' in obj === false)
 *
 * If options.strict is set, extraneous fields will be considered as invalid.
 * Example: validate({}, {a: 2}, {strict: true}) // false: I wasn't expecting a value in a
 */
register(function (definition, parse) {
	var extra

	if (typeof definition !== 'object' ||
		!definition ||
		(definition.constructor !== Object && Object.getPrototypeOf(definition) !== null)) {
		// Only match literal objects like {a: Number, ...} and null-prototype objects
		return
	}

	extra = {
		required: Object.create(null),
		optional: Object.create(null)
	}

	Object.keys(definition).forEach(function (key) {
		var optional = key[key.length - 1] === '?',
			newKey = optional ? key.substr(0, key.length - 1) : key,
			field = parse(definition[key])

		if (newKey in extra.optional || newKey in extra.required) {
			throw new Error('You can\'t define the same field twice: ' + key)
		}

		if (optional) {
			extra.optional[newKey] = field
		} else {
			extra.required[newKey] = field
		}
	})

	return extra
}, 'object', function (value, extra, options, path) {
	var key, field, subpath

	// Check required fields
	for (key in extra.required) {
		field = extra.required[key]
		subpath = path ? path + '.' + key : key
		if (!(key in value)) {
			throw new ValidationError('I was expecting a value', subpath)
		} else if (isEmpty(value[key])) {
			throw new ValidationError('I was expecting a non-empty value', subpath)
		}
		value[key] = field._validate(value[key], subpath, options)
	}

	// Check optional fields
	for (key in extra.optional) {
		field = extra.optional[key]
		subpath = path ? path + '.' + key : key
		if (key in value) {
			if (isEmpty(value[key])) {
				delete value[key]
			} else {
				value[key] = field._validate(value[key], subpath, options)
			}
		}
	}

	if (options.strict) {
		// Check for extraneous fields
		for (key in value) {
			subpath = path ? path + '.' + key : key
			if (!(key in extra.required) && !(key in extra.optional)) {
				throw new ValidationError('I wasn\'t expecting a value', subpath)
			}
		}
	}

	return value
}, function (extra) {
	var key, ret = Object.create(null)

	for (key in extra.required) {
		ret[key] = extra.required[key].toJSON()
	}
	for (key in extra.optional) {
		ret[key + '?'] = extra.optional[key].toJSON()
	}

	return ret
})

/**
 * Arrays
 * Example: [String] or {products: [{name: String, price: Number}]}
 * An empty array will not be tolerated, unless it is inside a hash map and marked as optional
 */
register(function (definition, parse) {
	if (!Array.isArray(definition) || definition.length !== 1) {
		// Only match arrays with exactly one element
		return
	}

	return parse(definition[0])
}, 'array', function (value, extra, options, path) {
	var i, subpath
	if (value.length === 0) {
		throw 'I was expecting a non-empty array'
	}
	for (i = 0; i < value.length; i++) {
		subpath = path ? path + '.' + i : i
		value[i] = extra._validate(value[i], subpath, options)
	}
	return value
}, function (extra) {
	return [extra.toJSON()]
})

/**
 * A number (double)
 * Example: {name: String, value: Number}
 */
register(Number, 'number', null, '$Number')

/**
 * A non-empty string (unless in a hash map, marked as optional)
 * Example: {name: {first: String, 'last?': String}}
 * The string will be HTML escaped if options.escape is true
 */
register(String, 'string', function (value, _, options) {
	if (value.length === 0) {
		throw 'I was expecting a non-empty string'
	}
	return options.escape ? escape(value) : value
}, '$String')

/**
 * A non-null generic object. No property is validated inside the object
 * Example: {date: Date, event: Object}
 */
register(Object, 'object', function (value) {
	if (value === null) {
		throw 'I was expecting a non-null object'
	}
}, '$Object')

/**
 * A non-empty generic array. No element is validated inside the array
 * Example: {arrayOfThings: Array}
 */
register(Array, 'array', function (value) {
	if (value.length === 0) {
		throw 'I was expecting a non-empty array'
	}
}, '$Array')

/** A bool value */
register(Boolean, 'boolean', null, '$Boolean')

/**
 * A date value in string format (like '2014-08-08T16:14:16.046Z')
 * It will convert the string into a date object
 * It accepts all formats defined by the Date(str) constructor, but the a ISO string is recommended
 */
register(Date, 'string', function (value) {
	var date = new Date(value)
	if (Number.isNaN(date.getTime())) {
		throw 'I was expecting a date string'
	}

	return date
}, '$Date')

/**
 * Any type
 */
register('*', '*', null, '*')

/**
 * A number range
 * Examples:
 * 'number(1,)'
 * 'number(,10)'
 * 'number(1,10)'
 */
registerTagged({
	tag: 'number',
	jsonType: 'number',
	minArgs: 2,
	maxArgs: 2,
	sparse: true,
	numeric: true
}, function (value, args) {
	if (args[0] !== undefined && value < args[0]) {
		throw 'I was expecting a value greater than ' + args[0]
	} else if (args[1] !== undefined && value > args[1]) {
		throw 'I was expecting a value less than ' + args[0]
	}
}, function (extra) {
	return extra.original
})

/**
 * A safe integer
 */
register('int', 'number', function (value) {
	if (!isSafeInt(value)) {
		throw 'I was expecting an integer'
	}
}, 'int')

/**
 * An integer range
 * Examples:
 * 'int(1,)'
 * 'int(,10)'
 * 'int(1,10)'
 */
registerTagged({
	tag: 'int',
	jsonType: 'number',
	minArgs: 2,
	maxArgs: 2,
	sparse: true,
	numeric: true
}, function (value, args) {
	if (!isSafeInt(value)) {
		throw 'I was expecting an integer'
	}

	if (args[0] !== undefined && value < args[0]) {
		throw 'I was expecting a value greater than ' + args[0]
	} else if (args[1] !== undefined && value > args[1]) {
		throw 'I was expecting a value less than ' + args[0]
	}
}, function (extra) {
	return extra.original
})

/**
 * A safe natural number
 */
register('uint', 'number', function (value) {
	if (!isSafeInt(value) || value < 0) {
		throw 'I was expecting a natural number'
	}
}, 'uint')

/**
 * An unsigned integer range
 * Examples:
 * 'uint(1,)'
 * 'uint(,10)'
 * 'uint(1,10)'
 */
registerTagged({
	tag: 'uint',
	jsonType: 'number',
	minArgs: 2,
	maxArgs: 2,
	sparse: true,
	numeric: true
}, function (value, args) {
	if (!isSafeInt(value) || value < 0) {
		throw 'I was expecting a natural number'
	}

	if (args[0] !== undefined && value < args[0]) {
		throw 'I was expecting a value greater than ' + args[0]
	} else if (args[1] !== undefined && value > args[1]) {
		throw 'I was expecting a value less than ' + args[0]
	}
}, function (extra) {
	return extra.original
})

/**
 * A non-empty string with limited length
 * Examples:
 * 'string(12)': exactly 12 chars
 * 'string(,100)': at most 100 chars
 * 'string(8,)': at least 8 chars
 * 'string(8,100)': at most 100, at least 8
 * The string will be HTML escaped if options.escape is true
 */
registerTagged({
	tag: 'string',
	jsonType: 'string',
	minArgs: 1,
	maxArgs: 2,
	sparse: true,
	numeric: true
}, function (value, args, options) {
	if (options.escape) {
		value = escape(value)
	}
	if (value.length === 0) {
		throw 'I was expecting a non-empty string'
	}

	if (args.length === 1) {
		// Exact length
		if (value.length !== args[0]) {
			throw 'I was expecting exactly ' + args[0] + ' chars'
		}
	} else {
		// Min/max length
		if (args[0] !== undefined && value.length < args[0]) {
			throw 'I was expecting at least ' + args[0] + ' chars'
		} else if (args[1] !== undefined && value.length > args[1]) {
			throw 'I was expecting at most ' + args[1] + ' chars'
		}
	}

	return value
}, function (extra) {
	return extra.original
})

/**
 * A non-empty hex-encoded string
 * Examples:
 * 'hex': a non-empty hex string
 * 'hex(12)': a hex-string with exactly 12 hex-chars (that is, 6 bytes)
 */
registerTagged({
	tag: 'hex',
	jsonType: 'string',
	maxArgs: 1,
	numeric: true
}, function (value, args) {
	if (value.length === 0) {
		throw 'I was expecting a non-empty string'
	} else if (!value.match(/^[0-9a-fA-F]+$/)) {
		throw 'I was expecting a hex-encoded string'
	} else if (args.length && value.length !== args[0]) {
		throw 'I was expecting a string with ' + args[0] + ' hex-chars'
	}
}, function (extra) {
	return extra.original
})

/**
 * A mongo objectId as a 24-hex-char string
 */
register('id', 'string', function (value) {
	if (!value.match(/^[0-9a-fA-F]{24}$/)) {
		throw 'I was expecting a mongo objectId'
	}
}, 'id')

/**
 * An email
 * Note: this regex isn't much strict on purpose, since:
 * 1) There is no complete regex that matches exactly the spec
 * 2) The right way to check if a email is valid is trying to send an message to it!
 *    (even this is broken: it's easy to get a temporary trash mailbox)
 */
register('email', 'string', function (value) {
	// The regex forces one '@', followed by at least one '.'
	if (!value.match(/^[^@]+@.+\.[^.]+$/i)) {
		throw 'I was expecting an email address'
	}
}, 'email')

/**
 * A string from the given set
 * Example: 'in(red, green, blue)' matches 'red', 'green' and 'blue', but no other
 */
registerTagged({
	tag: 'in',
	jsonType: 'string'
}, function (value, args) {
	if (args.indexOf(value) === -1) {
		throw 'I was expecting one of [' + args.join(', ') + ']'
	}
}, function (extra) {
	return extra.original
})

/**
 * A number from the given set
 * Example: 'numberIn(3, 14, 15)' matches 3, 14, 15, but no other
 */
registerTagged({
	tag: 'numberIn',
	jsonType: 'number',
	numeric: true
}, function (value, args) {
	if (args.indexOf(value) === -1) {
		throw 'I was expecting one of [' + args.join(', ') + ']'
	}
}, function (extra) {
	return extra.original
})

/**
 * A string that matches a custom regex
 * Example: /^\d+\.\d{2}$/ matches '3.14'
 */
register(function (definition) {
	if (definition instanceof RegExp) {
		return definition
	}
}, 'string', function (value, extra) {
	if (!value.match(extra)) {
		throw 'I was expecting a string that matches ' + extra
	}
}, function (extra) {
	var flags = (extra.global ? 'g' : '') +
		(extra.ignoreCase ? 'i' : '') +
		(extra.multiline ? 'm' : '')
	return '$RegExp:' + flags + ':' + extra.source
})

function isSafeInt(value) {
	return Number.isFinite(value) &&
		value > -9007199254740992 &&
		value < 9007199254740992 &&
		Math.round(value) === value
}